pertpy/tools/_distances/_distances.py

```diff
@@ -326,3 +393,151 @@ def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        raise NotImplementedError("CosineDistance cannot be called on a pairwise distance matrix.")
+
+
+class R2ScoreDistance(AbstractDistance):
+    """Coefficient of determination across genes between pseudobulk vectors."""
+
+    # NOTE: This is not a distance metric but a similarity metric.
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.accepts_precomputed = False
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        return r2_score(X.mean(axis=0), Y.mean(axis=0))
+
+    def from_precomputed(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        raise NotImplementedError("R2ScoreDistance cannot be called on a pairwise distance matrix.")
+
+
+class KLDivergence(AbstractDistance):
+    """Average of KL divergence between gene distributions of two groups
+
+    Assuming a Gaussian distribution for each gene in each group, calculates
+    the KL divergence between them and averages over all genes
+
+    """
+
+    # NOTE: This results in an approximation of KL-Divergence by calculating histogram over bins
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.accepts_precomputed = False
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        kl_all = []
+        for i in range(X.shape[1]):
+            x_mean, x_std = X[:, i].mean(), X[:, i].std()
+            y_mean, y_std = Y[:, i].mean(), Y[:, i].std()
+            kl = np.log(y_std / x_std) + (x_std**2 + (x_mean - y_mean) ** 2) / (2 * y_std**2) - 1 / 2
+            kl_all.append(kl)
+        return sum(kl_all) / len(kl_all)
+
+    def from_precomputed(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        raise NotImplementedError("KLDivergence cannot be called on a pairwise distance matrix.")
+
+
+class TTestDistance(AbstractDistance):
+    """Average of T test statistic between two groups assuming unequal variances"""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.accepts_precomputed = False
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, bins=10, **kwargs) -> float:
+        t_test_all = []
+        n1 = X.shape[0]
+        n2 = Y.shape[0]
+        for i in range(X.shape[1]):
+            m1, v1 = X[:, i].mean(), X[:, i].std() ** 2 * n1 / (n1 - 1)
+            m2, v2 = Y[:, i].mean(), Y[:, i].std() ** 2 * n2 / (n2 - 1)
+            vn1 = v1 / n1
+            vn2 = v2 / n2
+            t = (m1 - m2) / np.sqrt(vn1 + vn2)
+            t_test_all.append(abs(t))
+        return sum(t_test_all) / len(t_test_all)
+
+    def from_precomputed(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        raise NotImplementedError("TTestDistance cannot be called on a pairwise distance matrix.")
+
+
+class NBNLL(AbstractDistance):
+    """
+    Average of Negative Log likelihood (scalar) of group B cells
+    according to a NB distribution fitted over group A
```

Comment:
@mojtababahrami1993 Can you add here credit for the equation below to scvi authors? Although I did check this parameterization of the NLL equation myself, the code is technically from them. @Zethson Let us know if there's something else we should do here.

Reply:
If it's 100% copied we should add the scVI license here. If it's only adapted or something we can link to them.

Reply:
By adding the license I literally mean copying it into the folder where the Distance implementation lives. And state in the docstring nevertheless that you got it from scvi-tools

Reply:
If that's not too bad let's do that then. The formula is the general formula but even so there are different ways to call a gammaln, for example.